### PR TITLE
tag paginations & unkeep multiple for website

### PR DESCRIPTION
**Pull request by ahsu1230**
tag paginations support sorting (by name, by user's previous ordering, or by num_keeps)
unkeep multiple from a single library endpoint for website

@andrewconner

### CODE REVIEW

File: kifi-backend/modules/shoebox/test/com/keepit/controllers/website/LibraryControllerTest.scala [hunk 1]
```diff
@@ -962,5 +962,66 @@ class LibraryControllerTest extends Specification with ShoeboxTestInjector {
...
+        val k4Id: ExternalId[Keep] = ExternalId()
+        val json2 = Json.obj(
+          "ids" -> Json.toJson(Seq(k1.externalId, k2.externalId, k3.externalId, k4Id))
+        )
+        val request2 = FakeRequest("POST", testPathRemove).withBody(json2)
+        val result2 = libraryController.removeKeeps(pubId1)(request2)
+        status(result2) must equalTo(OK)
+        contentType(result2) must beSome("application/json")
+
+        Json.parse(contentAsString(result2)) must equalTo(Json.parse(
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/LibraryControllerTest.scala:1009]: @yipingliao 
usually for tests try to look for these `Json.parse(contentAsString(...))` for endpoint responses

File: kifi-backend/modules/shoebox/test/com/keepit/controllers/website/LibraryControllerTest.scala [hunk 2]
```diff
@@ -962,5 +962,66 @@ class LibraryControllerTest extends Specification with ShoeboxTestInjector {
...
+          keeps.length === 3
+          (keeps(0), keeps(1), keeps(2))
+        }
+
+        val testPathRemove = com.keepit.controllers.website.routes.LibraryController.removeKeeps(pubId1).url
+        val k4Id: ExternalId[Keep] = ExternalId()
+        val json2 = Json.obj(
+          "ids" -> Json.toJson(Seq(k1.externalId, k2.externalId, k3.externalId, k4Id))
+        )
+        val request2 = FakeRequest("POST", testPathRemove).withBody(json2)
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/LibraryControllerTest.scala]: @yipingliao 
look for `.withBody(json)` and then look for the json variable content to see REQUIRED input parameters for endpoints
optional parameters may not appear in tests, so then you'll have to look at the actual Controller source code

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala [hunk 3]
```diff
@@ -372,6 +372,15 @@ class KeepsController @Inject() (
     }
   }
 
+  def page(sort: String, offset: Int, pageSize: Int) = UserAction.async { request =>
+    val collectionsFuture = SafeFuture { collectionCommander.pageCollections(sort, offset, pageSize, request.userId) }
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala]: Why in a future?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 4]
```diff
@@ -371,6 +371,21 @@ class LibraryController @Inject() (
     }
   }
 
+  def removeKeeps(pubId: PublicId[Library]) = (UserAction andThen LibraryWriteAction(pubId))(parse.tolerantJson) { request =>
+    Library.decodePublicId(pubId) match {
+      case Failure(ex) =>
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: This is a bit weird, but at this point, we actually know the `pubId` can be decoded (because of  `LibraryWriteAction`). However, it makes for interesting looking code (ie, `val libId = Library.decodePublicId(pubId).get`), so keep this if you want (it just shouldn't ever run).

File: kifi-backend/modules/shoebox/conf/shoebox.routes [hunk 5]
```diff
@@ -296,6 +297,7 @@ POST    /site/libraries/:id/decline @com.keepit.controllers.website.LibraryContr
 POST    /site/libraries/:id/leave   @com.keepit.controllers.website.LibraryController.leaveLibrary(id: PublicId[Library])
 GET     /site/libraries/:id/keeps   @com.keepit.controllers.website.LibraryController.getKeeps(id: PublicId[Library], count: Int, offset: Int ?= 0, auth:Option[String] ?= None, passCode:Option[HashedPassPhrase] ?= None)
 POST    /site/libraries/:id/keeps   @com.keepit.controllers.website.LibraryController.addKeeps(id: PublicId[Library])
+DELETE  /site/libraries/:id/keeps   @com.keepit.controllers.website.LibraryController.removeKeeps(id: PublicId[Library])
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/conf/shoebox.routes]: Seems a bit strange to require a body for a DELETE method (especially when you're not deleting a "keeps" object, but a list of keeps in the body). The spec looks like it doesn't forbid it, but [rfc2616 4.3](http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.3) ("if the request method does not include defined semantics for an entity-body, then the message-body SHOULD be ignored when handling the request") says the server should ignore it.

Elsewhere, we use POST to an endpoint like `/site/libraries/:id/keeps/delete`

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/CollectionCommander.scala [hunk 6]
```diff
@@ -66,6 +66,26 @@ class CollectionCommander @Inject() (
     collections
   }
 
+  def pageCollections(sort: String, offset: Int, pageSize: Int, userId: Id[User]) = {
+    log.info(s"Getting all collections for $userId (sort $sort)")
+    val unsortedCollections = db.readOnlyMaster { implicit s =>
+      val colls = collectionRepo.getUnfortunatelyIncompleteTagSummariesByUser(userId)
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/CollectionCommander.scala]: `getUnfortunatelyIncompleteTagSummariesByUser` :). You likely want a separate repo method that isn't incomplete.

Also, below, `drop(...).take(...)` means they were all loaded into memory to begin with. Pagination was to avoid loading them into memory to begin with. As-is, this is basically doing what `tagService.js` does.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala [hunk 7]
```diff
@@ -372,6 +372,15 @@ class KeepsController @Inject() (
     }
   }
 
+  def page(sort: String, offset: Int, pageSize: Int) = UserAction.async { request =>
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala]: Naming? `KeepsController#page` doesn't seem to be related to tags.

Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala]: Hmm we don't have a `CollectionController` though =/

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala [hunk 8]
```diff
@@ -372,6 +372,15 @@ class KeepsController @Inject() (
...
+    val collectionsFuture = SafeFuture { collectionCommander.pageCollections(sort, offset, pageSize, request.userId) }
+    for {
+      collections <- collectionsFuture
+    } yield {
+      Ok(Json.obj("collections" -> collections))
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala]: Might be nice to kill this naming, since nothing client-side should reference a "collection" anymore (that naming is just baggage from tags being named "collections" at one point).

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/CollectionCommander.scala (hunk 6 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/CollectionCommander.scala]: Ok yea I was thinking about that - but if we want this to support sorting & pagination together, you kind of need to sort the entire list. Or can slick do the sorting for you?

File: kifi-backend/modules/shoebox/conf/shoebox.routes (hunk 5 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/conf/shoebox.routes]: Ok sounds good

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/CollectionCommander.scala (hunk 6 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/CollectionCommander.scala]: Yeah, you can do it in the database. Sorting by number of keeps may be a bit expensive if you have quite a bit, but a _lot_ cheaper than iterating for each tag and fetching the count.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 4 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: I see :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/CollectionCommander.scala (hunk 6 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/CollectionCommander.scala]: It's not too bad. For our worst user, the query is less than a second:

``` sql
select c.*,count(kc.id) as keep_count from collection c, keep_to_collection kc where kc.collection_id = c.id and user_id=<userid> and kc.state = "active" and c.state="active" group by c.id order by keep_count desc limit 10;
```

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala (hunk 8 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/KeepsController.scala]: ah okay good point.

File: kifi-backend/modules/shoebox/conf/shoebox.routes (hunk 5 above)
Comment (2is10) [inline kifi-backend/modules/shoebox/conf/shoebox.routes]: Yeah, URL should fully specify what’s being deleted when using the DELETE method.

File: kifi-backend/modules/shoebox/test/com/keepit/controllers/website/LibraryControllerTest.scala [hunk 9]
```diff
@@ -962,5 +962,66 @@ class LibraryControllerTest extends Specification with ShoeboxTestInjector {
...
+        val keepsToAdd =
+          RawBookmarkRepresentation(title = Some("title 11"), url = "http://www.hi.com11", isPrivate = None) ::
+            RawBookmarkRepresentation(title = Some("title 21"), url = "http://www.hi.com21", isPrivate = None) ::
+            RawBookmarkRepresentation(title = Some("title 31"), url = "http://www.hi.com31", isPrivate = None) ::
+            Nil
+
+        inject[FakeUserActionsHelper].setUser(user1)
+
+        val json = Json.obj(
+          "keeps" -> JsArray(keepsToAdd map { k => Json.toJson(k) })
```
Comment (2is10) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/LibraryControllerTest.scala]: By the way, this can be shortened to:
`"keeps" -> keepsToAdd`
With `Json.obj`, the coercion to `JsValue` types is automatic.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 10]
```diff
@@ -371,6 +371,21 @@ class LibraryController @Inject() (
...
+        BadRequest(Json.obj("error" -> "invalid_id"))
+      case Success(libId) =>
+        val keepExtIds = (request.body \ "ids").as[Seq[ExternalId[Keep]]]
+        val source = KeepSource.site
+        implicit val context = heimdalContextBuilder.withRequestInfoAndSource(request, source).build
+        keepsCommander.unkeepManyFromLibrary(keepExtIds, libId, request.userId) match {
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: The `keepsCommander` also has a `unkeepOneFromLibrary` function. Should we also expose that on the controller? Currently there is a `/site/keeps/:id/delete` endpoint that calls `KeepsController.unkeep()`, but I am not sure if that does what it's supposed to do or whether going forward we should use `unkeepOneFromLibrary` anyway. Can you look into that? Thanks!

Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: You're right. This `unkeep()` will only unkeep the first instance of a keep externalId & userId. While it should still work (even if the parameters are not library aware), but we should probably be more explicit & make things library aware. But... should we replace the endpoint now?

Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: This is a good point. I think we should use simple endpoints for simple actions (which are more common) whenever possible.

@ahsu1230 Using the `DELETE` method is appropriate for deleting a single keep, with the keep ID in the URL.

File: kifi-backend/modules/shoebox/conf/shoebox.routes [hunk 11]
```diff
@@ -296,6 +297,8 @@ POST    /site/libraries/:id/decline @com.keepit.controllers.website.LibraryContr
 POST    /site/libraries/:id/leave   @com.keepit.controllers.website.LibraryController.leaveLibrary(id: PublicId[Library])
 GET     /site/libraries/:id/keeps   @com.keepit.controllers.website.LibraryController.getKeeps(id: PublicId[Library], count: Int, offset: Int ?= 0, auth:Option[String] ?= None, passCode:Option[HashedPassPhrase] ?= None)
 POST    /site/libraries/:id/keeps   @com.keepit.controllers.website.LibraryController.addKeeps(id: PublicId[Library])
+POST    /site/libraries/:id/keeps/delete   @com.keepit.controllers.website.LibraryController.removeKeeps(id: PublicId[Library])
+DELETE  /site/libraries/:id/keeps/:keepId   @com.keepit.controllers.website.LibraryController.removeKeep(id: PublicId[Library], keepId: ExternalId[Keep])
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/conf/shoebox.routes:301]: unkeep one from a library endpoint provided (with test) @yipingliao @2is10 
does this work?

Comment (yipingliao) [inline kifi-backend/modules/shoebox/conf/shoebox.routes:301]: I thought that per @2is10 we don't need these, if we're sure that the current keeps delete endpoint works correctly w/r to libraries?

**Comment by 2is10** [comment]
:+1: great, thanks

File: kifi-backend/modules/shoebox/conf/shoebox.routes (hunk 11 above)
Comment (2is10) [inline kifi-backend/modules/shoebox/conf/shoebox.routes:301]: It’s better to specify the library ID in the URL.

Comment (ahsu1230) [inline kifi-backend/modules/shoebox/conf/shoebox.routes:301]: Oh I was under the impression that because the original unkeeping doesn't clearly specify which library it is unkeeping from, we should add it in and use this new endpoint instead.
Down the road (maybe after our release) we should be able to clean this API up and consolidate the keeping/unkeeping functions

edit: ok looks like we're in agreement lol

Comment (andrewconner) [inline kifi-backend/modules/shoebox/conf/shoebox.routes:301]: I agree with Jared, moving to new APIs that specify libraries is better.

Comment (andrewconner) [inline kifi-backend/modules/shoebox/conf/shoebox.routes:301]: (ie, this is a good change)
